pkg/binding/bindingpolicy-resolution-broker.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package binding
+
+import (
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/abstract"
+	"github.com/kubestellar/kubestellar/pkg/util"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sync"
+)
+
+// ResolutionBroker allows for the registration of callback functions that
+// are called when a resolution is updated.
+// It also allows for the retrieval of resolutions for a given bindingPolicyKey.
+type ResolutionBroker interface {
+	// RegisterCallback adds a new callback function that will be called,
+	// on a separate goroutine, when a resolution is updated. The callback
+	// function should accept a bindingPolicyKey.
+	// There is no deduplication of callbacks.
+	RegisterCallback(func(bindingPolicyKey string))
+
+	// GetResolution retrieves the resolution for a given bindingPolicyKey.
+	// If no resolution is associated with the given key, nil is returned.
+	GetResolution(bindingPolicyKey string) *Resolution
+}
+
+// Resolution is a struct that represents the resolution of a binding policy.
+// It contains the destinations and object data for the resolution.
+type Resolution struct {
```

Comment:
Is it necessary to make this an immutable data structure? What would go wrong if it were an interface that provides procedural acces? Something like the following?

```
type Resolution interface {
  Destinations(yield(v1alpha1.Destination) bool)
  Workload(yield(util.ObjectIdentifier, struct{ResourceVersion  string; StatusCollectors /*not modified during this call*/ Set[string]}) bool)
}
```